Add GLSL-based noise(vec2) function to p5.strands (#7897)

LalitNarayanYadav opened the pull request:
### Summary

This PR implements a `noise(vec2)` function in `p5.strands` to enable shader-compatible noise effects. It addresses [#7897](https://github.com/processing/p5.js/issues/7897), which is part of the broader effort to bridge core p5.js functions with p5.strands ([#7849](https://github.com/processing/p5.js/issues/7849)).

---

### Changes Made

- Created `src/webgl/shaders/functions/noise.glsl.js` with a basic fractal noise implementation (3 octaves).
- Registered `noise(vec2)` in `builtInGLSLFunctions` in `ShaderGenerator.js` with `isp5Function: false`.
- Injected the GLSL noise code into all fragment shaders via `GLOBAL_SHADER.output.fragmentDeclarations.add(noiseGLSL)`.

#### PR Checklist

- [x] `npm run lint` passes
- [x] [Inline reference] is included / updated
- [x] [Unit tests] are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

RandomGamingDev commented:
Looks good to me :D

LalitNarayanYadav commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1632,6 +1634,7 @@ function shadergenerator(p5, fn) {
 }
   
   
+GLOBAL_SHADER.output.fragmentDeclarations.add(noiseGLSL);
```
Thanks @davepagurek! That makes sense. I’ve moved the `noiseGLSL` injection into a custom `fn.noise` function wrapper, similar to what was done for `lerp`. That way it’s only added during shader generation and avoids undefined `GLOBAL_SHADER`.

Also working on a test sketch under `src/preview/global/sketch.js` as suggested. Appreciate the review!

perminder-17 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1629,10 +1631,19 @@ function shadergenerator(p5, fn) {
   };
+  fn.noise = function (...args) {
+    if (GLOBAL_SHADER?.isGenerating) {
+      GLOBAL_SHADER.output.fragmentDeclarations.add(noiseGLSL);
```
Would it be worth adding a vertex declarations for positions before fragment declarations?
```js
GLOBAL_SHADER.output.vertexDeclarations.add(noiseGLSL);
```

perminder-17 reviewed (changes requested):
Great work on this so far, just some minor thoughts and would be good to.

davepagurek commented on src/webgl/shaders/functions/noise.glsl.js:
```diff
@@ -0,0 +1,15 @@
+export default /* glsl */ `
```
good catch @perminder-17!

LalitNarayanYadav replied:
Thanks a lot for pointing that out @perminder-17 and @davepagurek ! I've renamed the file to `noiseGLSL.glsl` and updated the import accordingly to avoid the Rollup plugin conflict. Let me know if anything else needs adjusting!

perminder-17 commented on src/webgl/shaders/functions/noiseGLSL.glsl:
```diff
@@ -0,0 +1,13 @@
+float baseNoise(vec2 st) {
```
Hi @LalitNarayanYadav while I tested it locally, I found that the things are little jittery and we want the noise to be smoother, maybe we just need a different basis for each octave.

e.g. something like the examples here, but, maybe finding where they came from to double check the licenses, or doing something similar but different https://gist.github.com/patriciogonzalezvivo/670c22f3966e662d2f83

davepagurek commented on src/webgl/shaders/functions/noiseGLSL.glsl:
```diff
@@ -0,0 +1,13 @@
+float baseNoise(vec2 st) {
```
off the top of my head I think we were considering at least `random()`, so I think we can keep this folder structure for that.

LalitNarayanYadav commented:
Thanks @perminder-17 and @davepagurek for the feedback!

I've updated the GLSL noise to use a smoother implementation based on [this MIT-licensed source](https://gist.github.com/patriciogonzalezvivo/670c22f3966e662df83), and added fallback support for core `p5.js` noise() when called outside of strands, similar to how lerp is handled. Let me know what you think!

perminder-17 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1578,6 +1579,7 @@ function shadergenerator(p5, fn) {
     ],
     'sqrt': { args: ['genType'], returnType: 'genType', isp5Function: true},
     'step': { args: ['genType', 'genType'], returnType: 'genType', isp5Function: false},
+    'noise': { args: ['vec2'], returnType: 'float', isp5Function: false },
```
I think in the glsl code, we have a `baseNoise()` function returning float. So, either you have to rename at the js file or in the glsl file. Probably changing from `baseNoise()` to `noise()` in the glsl is what I would prefer.

perminder-17 commented on src/webgl/shaders/functions/noiseGLSL.glsl:
```diff
@@ -0,0 +1,29 @@
+  return -1.0 + 2.0 * fract(sin(st) * 43758.5453123);
+}
+
+float baseNoise(vec2 st) {
```
```suggestion
float noise(vec2 st) {
```
